flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/resources/ClusterResourceManager.java

```diff
@@ -0,0 +1,282 @@
+ * resources.
+ *
+ * <p>The goal is to provide a good indicator for whether resources needed for autoscaling are going
+ * to be available. This is achieved by pulling the node resource usage from the Kubernetes cluster
+ * at a regular configurable interval, after which we use this data to simulate adding / removing
+ * resources (pods). Note that this is merely a (pretty good) heuristic because the Kubernetes
+ * scheduler has the final saying. However, we prevent 99% of the scenarios after pipeline outages
+ * which can lead to massive scale up where all pipelines may be scaled up at the same time and
+ * exhaust the number of available resources.
+ *
+ * <p>The simulation can run on a fixed set of Kubernetes nodes. Additionally, if we detect that the
+ * cluster is using the Kubernetes Cluster Autoscaler, we will use this data to extrapolate the
+ * number of nodes to the maximum defined nodes in the autoscaler configuration.
+ *
+ * <p>We currently track CPU and memory. Ephemeral storage is missing because there is no easy way
+ * to get node statics on free storage.
+ */
+public class ClusterResourceManager implements ResourceCheck {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ClusterResourceManager.class);
+
+    /** ConfigMap name of the Kubernetes Cluster Autoscaler. */
+    static final String CLUSTER_AUTOSCALER_CONFIG_MAP = "cluster-autoscaler-status";
+
+    /** EKS specific node group information. Code still works without this label. */
+    static final String LABEL_NODE_GROUP = "eks.amazonaws.com/nodegroup";
+
+    private final Duration refreshInterval;
+    private final KubernetesClient kubernetesClient;
+
+    @VisibleForTesting ClusterResourceView clusterResourceView;
+
+    public ClusterResourceManager(Duration refreshInterval, KubernetesClient kubernetesClient) {
+        this.refreshInterval = refreshInterval;
+        this.kubernetesClient = kubernetesClient;
+    }
+
+    @Override
+    public synchronized boolean trySchedule(
+            int currentInstances,
+            int newInstances,
+            double cpuPerInstance,
+            double memoryPerInstance) {
+
+        if (refreshInterval.isNegative()) {
+            // Feature disabled
+            return true;
+        }
+
+        if (shouldRefreshView(clusterResourceView, refreshInterval)) {
+            try {
+                clusterResourceView = createResourceView(kubernetesClient);
+            } catch (KubernetesClientException e) {
+                if (e.getCode() == 403) {
+                    LOG.warn(
+                            "No permission to retrieve node resource usage. Resource check disabled.");
+                    return true;
+                }
+                throw e;
+            }
+        }
+
+        if (newInstances <= currentInstances) {
+            LOG.debug("Fewer or same amount of resources used after scaling.");
+            return true;
+        }
```

Comment:
This should be moved in front of everything

Reply:
That's what I also thought but moving this in front of everything will corrupt the resource view because we will never remove resources when we scale down.